File: experiments/runs/2026-09-14_17-11-19_claim-office-example-mapping_external-tcr-kentbeck-2026-09-14-cc_opus-5-requesty-no-thinking/src/policy.ts
import { isComponent, lookup } from './catalog.js';

export interface Item {
  type: string;
  material?: string;
  enchantment?: number;
  cursed?: boolean;
}

const BLOCK_SIZE = 3;
const BLOCK_PREMIUM = 60;

export class UnknownItemError extends Error {
  constructor(type: string) {
    super(`unknown item type: ${type}`);
  }
}

function specOf(item: Item) {
  const spec = lookup(item.type);
  if (!spec) throw new UnknownItemError(item.type);
  return spec;
}

export function insuranceSum(items: Item[]): number {
  return items.reduce((sum, item) => sum + specOf(item).insuranceValue, 0);
}

/** Base premium of a single item, ignoring component block discounts. */
export function itemBasePremium(item: Item): number {
  return specOf(item).basePremium;
}

/** Base premium of the whole policy, applying the block of 3 alike components. */
export function policyBasePremium(items: Item[]): number {
  let total = 0;
  const componentCounts = new Map<string, number>();
  for (const item of items) {
    const spec = specOf(item);
    if (isComponent(item.type)) {
      componentCounts.set(item.type, (componentCounts.get(item.type) ?? 0) + 1);
    } else {
      total += spec.basePremium;
    }
  }
  for (const [type, count] of componentCounts) {
    const spec = lookup(type)!;
    total += count === BLOCK_SIZE ? BLOCK_PREMIUM : count * spec.basePremium;
  }
  return total;
}

File: experiments/runs/2026-09-20_07-28-11_claim-office-java-example-mapping_baseline-inline-tdd-v1-cc_opus-5-no-thinking/src/main/java/Policy.java
import java.math.BigDecimal;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Premium calculation for a list of insured items. */
public final class Policy {

    private Policy() {
    }

    /**
     * Sum of the item base premiums. Components of the same type form a building
     * block at a special rate when there are exactly {@link PriceList#BLOCK_SIZE}
     * of them; any other count is charged per component.
     */
    public static BigDecimal basePremium(List<Item> items) {
        BigDecimal total = BigDecimal.ZERO;
        Map<String, Integer> componentCounts = new LinkedHashMap<>();

        for (Item item : items) {
            String type = item.type();
            if (PriceList.isComponent(type)) {
                componentCounts.merge(type, 1, Integer::sum);
            } else {
                total = total.add(PriceList.basePremium(type));
            }
        }

        for (Map.Entry<String, Integer> entry : componentCounts.entrySet()) {
            total = total.add(componentsPremium(entry.getKey(), entry.getValue()));
        }
        return total;
    }

    private static BigDecimal componentsPremium(String type, int count) {
        if (count == PriceList.BLOCK_SIZE) {
            return PriceList.BLOCK_PREMIUM;
        }
        return PriceList.basePremium(type).multiply(new BigDecimal(count));
    }
}
